### PR TITLE
[MLIR] Add f8E4M3 IEEE 754 type

### PR DESCRIPTION
**Pull request by apivovarov**
This PR adds `f8E4M3` type to mlir.

`f8E4M3` type  follows IEEE 754 convention

```c
f8E4M3 (IEEE 754)
- Exponent bias: 7
- Maximum stored exponent value: 14 (binary 1110)
- Maximum unbiased exponent value: 14 - 7 = 7
- Minimum stored exponent value: 1 (binary 0001)
- Minimum unbiased exponent value: 1 − 7 = −6
- Precision specifies the total number of bits used for the significand (mantisa), 
    including implicit leading integer bit = 3 + 1 = 4
- Follows IEEE 754 conventions for representation of special values
- Has Positive and Negative zero
- Has Positive and Negative infinity
- Has NaNs

Additional details:
- Max exp (unbiased): 7
- Min exp (unbiased): -6
- Infinities (+/-): S.1111.000
- Zeros (+/-): S.0000.000
- NaNs: S.1111.{001, 010, 011, 100, 101, 110, 111}
- Max normal number: S.1110.111 = +/-2^(7) x (1 + 0.875) = +/-240
- Min normal number: S.0001.000 = +/-2^(-6)
- Max subnormal number: S.0000.111 = +/-2^(-6) x 0.875 = +/-2^(-9) x 7
- Min subnormal number: S.0000.001 = +/-2^(-6) x 0.125 = +/-2^(-9)
```

Related PRs:
- [PR-97179](https://github.com/llvm/llvm-project/pull/97179) [APFloat] Add support for f8E4M3 IEEE 754 type
- jax-ml/ml_dtypes [PR-161](https://github.com/jax-ml/ml_dtypes/pull/161) Add float8_e4m3

### CODE REVIEW

**Comment by apivovarov** [comment]
Reed, Benjamin, Can you look at this PR too? @reedwm @d0k

**Comment by makslevental** [comment]
Ah so this is the reason for the previous PR :)

**Comment by tschuett** [comment]
If this is a new float type, could you please split out the apfloat changes in separate PR.

**Comment by apivovarov** [comment]
> If this is a new float type, could you please split out the apfloat changes in separate PR.

Sure - [PR-97179](https://github.com/llvm/llvm-project/pull/97179) Add f8E4M3 IEEE 754 type to llvm

**Comment by makslevental** [comment]
Hey @apivovarov I think this needs a rebas?

**Comment by apivovarov** [comment]
> Hey @apivovarov I think this needs a rebase on top #97179?

Just rebased

**Comment by makslevental** [comment]
> Just rebased and changed the commit message to match new PR title - `[MLIR] Add f8E4M3 IEEE 754 type`

Ah I forgot you need actually change the base but just rebase to get the commits in the other PR to disappear (which is what I wanted for ease of review). No big deal (I can hide the first commit still in the UI). I'll review soon (either tonight or tomorrow morning).

**Comment by makslevental** [comment]
Hey @apivovarov I just wanted to say my bad for not getting to this yet - I'm buried at work but I haven't forgotten. If it's hipri I can find someone else to review. Otherwise ~couple of days I think and I'll be clear to review.

**Comment by apivovarov** [comment]
> Hey @apivovarov I just wanted to say my bad for not getting to this yet - I'm buried at work but I haven't forgotten. If it's hipri I can find someone else to review. Otherwise ~couple of days I think and I'll be clear to review.

Hi Maksim, I think it has standard priority.  APFloat PR is quiet for the last 4 days... https://github.com/llvm/llvm-project/pull/97179 . 4th of July week...

File: llvm/lib/Support/APFloat.cpp
Comment (makslevental) [inline llvm/lib/Support/APFloat.cpp:139]: Just sanity checking: this is 

```
maxExponent = 7
minExponent = -6
precision = 4
sizeInBits = 8
```
?

File: clang/include/clang/AST/Stmt.h
Comment (makslevental) [inline clang/include/clang/AST/Stmt.h:466]: Didn't I see this in another PR you sent? If not, is there a way to separate this into PRs for LLVM/Clang and PRs for MLIR?

File: llvm/unittests/ADT/APFloatTest.cpp
Comment (makslevental) [inline llvm/unittests/ADT/APFloatTest.cpp:6551]: Sanity check: this test is basically exactly the same as above except with exponent adjusted?

**Review by makslevental — COMMENTED** [review]
Man talk about boiler plate 😄. 

Anyway this all looks basically fine. Except for the one ask about splitting this up I think I can approve (modulo the tests passing). I'll double check tomorrow on a bigger screen (currently on my laptop) so I can make sure everything lines up with the copy-pasted tests.

File: llvm/lib/Support/APFloat.cpp
Comment (apivovarov) [reply llvm/lib/Support/APFloat.cpp:139]: Maksim, Adding f8E4M3 IEE754 dtype was split into two PRs:

1. llvm/clang/APFloat changes [PR-97179](https://github.com/llvm/llvm-project/pull/97179) 
2. MLIR changes [PR-97118](https://github.com/llvm/llvm-project/pull/97118)  (this PR)

PR-97118 depends on PR-97179. 

Currently you see two commits in PR-97118.

Once PR-97179 is merged - I'll rebase PR-97118 and it will contain just one commit - just MLIR changes.

Can you review just MLIR changes in this PR (PR-97118). All comments for llvm/clang/APFloat should go to [PR-97179](https://github.com/llvm/llvm-project/pull/97179).

File: clang/include/clang/AST/Stmt.h
Comment (apivovarov) [reply clang/include/clang/AST/Stmt.h:466]: This code is from [PR-97179](https://github.com/llvm/llvm-project/pull/97179). Need to use one additional byte to store new item with id 16 (total 17 items)

File: llvm/unittests/ADT/APFloatTest.cpp
Comment (apivovarov) [reply llvm/unittests/ADT/APFloatTest.cpp:6551]: llvm/clang/APFloat code is from [PR-97179](https://github.com/llvm/llvm-project/pull/97179).

File: llvm/lib/Support/APFloat.cpp
Comment (apivovarov) [reply llvm/lib/Support/APFloat.cpp:139]: `f8E4M3` type follows IEEE 754 convention:
- Exponent bias: 7
- Maximum stored exponent value: 14 (binary 1110)
- Maximum unbiased exponent value: 14 - 7 = 7
- Minimum stored exponent value: 1 (binary 0001)
- Minimum unbiased exponent value: 1 − 7 = −6
- Precision specifies the total number of bits used for the significand (mantisa), including implicit leading integer bit = 3 + 1 = 4

`fltSemantics semFloat8E4M3 = {7, -6, 4, 8};`:
- maxExponent = 7
- minExponent = -6
- precision = 4
- sizeInBits = 8
- nonFiniteBehavior = fltNonfiniteBehavior::IEEE754 (default)
- nanEncoding = fltNanEncoding::IEEE (default)


The implementation patterns for `f8E4M3` should be similar to those of the existing `f8E5M2` type.

File: llvm/unittests/ADT/APFloatTest.cpp
Comment (apivovarov) [reply llvm/unittests/ADT/APFloatTest.cpp:6551]: f8E4M3 Min normal number: S.0001.000 = +/-2^(-6)
```
APFloat::getSmallestNormalized(APFloat::Float8E4M3(), false) // returns 0.0001.000 => +2^(-6)
APFloat::getSmallestNormalized(APFloat::Float8E4M3(), true) // returns 1.0001.000 => -2^(-6)
```

**Comment by apivovarov** [comment]
@makslevental, the APFloat PR-97179 has been merged recently. Now, this PR-97118 only contains the MLIR part. Could you please review it once more?

**Review by makslevental — APPROVED** [review]
Good job - very careful boilerplate :) Sorry it took so long to approve :(

**Pull request merged**
